crates/zizmor/src/audit/ref_version_mismatch.rs

```diff
@@ -0,0 +1,130 @@
+        Regex::new(r"#\s*(?:version|ver)\s*[:=]\s*(v?\d+(?:\.\d+)*(?:\.\d+)?)").unwrap(),
+    ]
+});
+
+impl RefVersionMismatch {
+    fn extract_version_from_comments<'doc>(
+        &self,
+        comments: &'doc Vec<Comment<'doc>>,
+    ) -> Option<&'doc str> {
+        for comment in comments {
+            for pattern in VERSION_COMMENT_PATTERNS.iter() {
+                if let Some(captures) = pattern.captures(comment.as_ref()) {
+                    if let Some(version_match) = captures.get(1) {
+                        return Some(version_match.as_str());
+                    }
+                }
+            }
+        }
+        None
+    }
+}
+
+impl Audit for RefVersionMismatch {
+    fn new(state: &AuditState<'_>) -> Result<Self, AuditLoadError> {
+        if state.no_online_audits {
+            return Err(AuditLoadError::Skip(anyhow!(
+                "offline audits only requested"
+            )));
+        }
+
+        let Some(client) = state.github_client() else {
+            return Err(AuditLoadError::Skip(anyhow!(
+                "can't run without a GitHub API token"
+            )));
+        };
+
+        Ok(Self { client })
+    }
+
+    fn audit_step<'doc>(&self, step: &Step<'doc>) -> anyhow::Result<Vec<Finding<'doc>>> {
+        let mut findings = vec![];
+
+        let Some(Uses::Repository(uses)) = step.uses() else {
+            return Ok(findings);
+        };
+
+        // Only check steps that have commit refs (not symbolic refs like v1.0.0)
+        let Some(commit_sha) = uses.commit_ref() else {
+            return Ok(findings);
+        };
+
+        // For now, we'll search the entire document source for version patterns
+        // that might correspond to this uses line. This is a simplified approach.
+        let step_location = step.location();
+        let uses_location = step_location
+            .with_keys(&["uses".into()])
+            .concretize(step.document())?;
+
+        let Some(version_from_comment) =
+            self.extract_version_from_comments(&uses_location.concrete.comments)
+        else {
+            return Ok(findings);
+        };
+
+        let Some(commit_for_ref) =
+            self.client
+                .commit_for_ref(&uses.owner, &uses.repo, version_from_comment)?
+        else {
+            return Ok(findings);
+        };
+
+        if commit_for_ref != commit_sha {
+            let mut annotation =
+                format!("version comment {version_from_comment:?} resolves to {commit_for_ref:?}");
+            if let Some(suggestion) =
+                self.client
+                    .longest_tag_for_commit(&uses.owner, &uses.repo, commit_sha)?
+            {
+                annotation.push_str(&format!(
+                    ", try updating the comment to {:?}",
+                    suggestion.name
+                ));
+            }
+            findings.push(
+                Self::finding()
+                    .severity(Severity::Low)
```

Comment:
I could see an argument for this being medium -- it breaks tools like Dependabot and makes people think they're up to date/using a specific version, but actually aren't. OTOH I don't feel super strong about it; thoughts?